File: aoc15/day_10/day_10.ml
let input =
  List.map (fun c -> Char.code c - Char.code '0') (Input.chars "input")
in

let rec generate lst num cnt res =
  match lst with
  | [] -> num :: cnt :: res
  | x :: rs ->
      if x = num then generate rs num (cnt + 1) res
      else generate rs x 1 (num :: cnt :: res)
in

let apply input n =
  let digits = ref input in
  for _ = 1 to n do
    digits := List.rev (generate !digits (List.nth !digits 0) 0 [])
  done;
  List.length !digits
in

let first = apply input 40 in
Printf.printf "First: %d\n" first;

let second = apply input 50 in
Printf.printf "Second: %d\n" second

File: aoc15/go/day_10/day_10.go
package main

import (
	"fmt"
	"os"
)

func apply(nums []byte) []byte {
	res := make([]byte, 0, 2*len(nums))
	for i := 0; i < len(nums); {
		n := 0
		for i+n < len(nums) && nums[i+n] == nums[i] {
			n++
		}
		res = append(res, byte(n), nums[i])
		i += n
	}
	return res
}

func run(initNums []byte, iters int) int {
	nums := initNums
	for i := 0; i < iters; i++ {
		nums = apply(nums)
	}
	return len(nums)
}

func main() {
	input, _ := os.ReadFile("input")

	nums := make([]byte, len(input)-1)
	for i, c := range input {
		if c >= '0' && c <= '9' {
			nums[i] = c - '0'
		}
	}

	fmt.Println("First:", run(nums, 40))
	fmt.Println("Second:", run(nums, 50))
}
